impl/maven-impl/src/test/java/org/apache/maven/impl/PathSelectorTest.java

```diff
@@ -0,0 +1,92 @@
+import org.junit.jupiter.api.io.TempDir;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class PathSelectorTest {
+    /**
+     * Creates a temporary directory and checks its list of content based on patterns.
+     *
+     * @param directory temporary directory where to create a tree
+     * @throws IOException if an error occurred while creating a temporary file or directory
+     */
+    @Test
+    public void testTree(final @TempDir Path directory) throws IOException {
+        Path foo = Files.createDirectory(directory.resolve("foo"));
+        Path bar = Files.createDirectory(foo.resolve("bar"));
+        Path baz = Files.createDirectory(directory.resolve("baz"));
+        Files.createFile(directory.resolve("root.txt"));
+        Files.createFile(bar.resolve("leaf.txt"));
+        Files.createFile(baz.resolve("excluded.txt"));
+        assertFilteredFilesContains(directory, "", "root.txt", "foo/bar/leaf.txt");
+        assertFilteredFilesContains(directory, "glob:", "foo/bar/leaf.txt");
+    }
+
+    /**
+     * Asserts that the filtered set of paths contains the given items and nothing more.
+     *
+     * @param directory the temporary directory containing the files to test
+     * @param syntax syntax to test, either an empty string of {@code "glob:"}
+     * @param expected the expected paths
+     * @throws IOException if an error occurred while listing the files
+     */
+    private static void assertFilteredFilesContains(final Path directory, final String syntax, final String... expected)
+            throws IOException {
+        var includes = List.of(syntax + "**/*.txt");
+        var excludes = List.of(syntax + "baz/**");
+        var matcher = new PathSelector(directory, includes, excludes, false);
+        Set<Path> filtered =
+                new HashSet<>(Files.walk(directory).filter(matcher::matches).toList());
+        for (String path : expected) {
+            assertTrue(filtered.remove(directory.resolve(path)), path);
+        }
+        assertTrue(filtered.isEmpty(), filtered.toString());
+    }
+
+    /**
+     * Tests the omission of unnecessary excludes.
+     *
+     * Note: at the time of writing this test (April 2025), the list of excludes goes down from 40 to 17 elements.
+     * This is not bad, but we could do better with, for example, a special treatment of the excludes that are
+     * for excluding an entire directory.
+     */
+    @Test
+    public void testExcludeOmission() {
+        Path directory = Path.of("dummy");
+        var includes = List.of("**/*.java");
+        var excludes = List.of("baz/**");
+        var matcher = new PathSelector(directory, includes, excludes, true);
+        String s = matcher.toString();
+        assertTrue(s.contains("glob:**/*.java"));
+        assertFalse(s.contains("project.pj")); // Unnecessary exclusion should have been omitted.
+        assertFalse(s.contains(".DS_Store"));
```

Comment:
```suggestion
        assertThat(s).contains(".DS_Store");
```
new code, already, considered, outdated - therefore increasing technical dept.

to reach next lvl, we need leadership and team up to standardize `DOD`.

<img width="989" alt="image" src="https://github.com/user-attachments/assets/ee0bf423-29b4-44e0-8d0b-36f9cbef9fcd" />

consider migration to avoid boilerplate API usage.

- https://github.com/apache/maven/pull/2307
- https://docs.openrewrite.org/recipes/java/testing/assertj/junittoassertj

Reply:
Let's do the refactoring in the other PR related to AssertJ migration.

Reply:
yes

Reply:
this will serve our need:

https://docs.openrewrite.org/recipes/java/testing/junit5/junit5bestpractices

<img width="1525" alt="image" src="https://github.com/user-attachments/assets/0cd8121c-ba09-44af-b408-604711e7817b" />

- https://github.com/apache/maven/pull/2320